server/clients/clients.go

```diff
@@ -49,29 +49,17 @@ func Deactivate(
 	db database.Database,
 	refKey types.ClientRefKey,
 ) (*database.ClientInfo, error) {
-	clientInfo, err := db.FindClientInfoByRefKey(ctx, refKey)
+	// TODO(hackerwins): We need to remove the presence of the client from the document.
+	// Be careful that housekeeping is executed by the leader. And documents are sharded
+	// by the servers in the cluster. So, we need to consider the case where the leader is
+	// not the same as the server that handles the document.
+
+	clientInfo, err := db.DeactivateClient(ctx, refKey)
 	if err != nil {
 		return nil, err
 	}
 
 	for docID, clientDocInfo := range clientInfo.Documents {
-		isAttached, err := clientInfo.IsAttached(docID)
-		if err != nil {
-			return nil, err
-		}
-		if !isAttached {
-			continue
-		}
-
-		if err := clientInfo.DetachDocument(docID); err != nil {
-			return nil, err
-		}
-
-		// TODO(hackerwins): We need to remove the presence of the client from the document.
-		// Be careful that housekeeping is executed by the leader. And documents are sharded
-		// by the servers in the cluster. So, we need to consider the case where the leader is
-		// not the same as the server that handles the document.
-
 		if err := db.UpdateSyncedSeq(
```

Comment:
It would be good to add two comments to this logic.

A. When deactivating a client, we need to update three properties(`ClientInfo.Status`, `ClientInfo.Documents` and `SyncedSeq`) in DB. Updating the sub-properties of `ClientInfo` guarantees atomicity as it involves a single MongoDB document. However, `SyncedSeq`s are stored in separate documents, so we can't ensure atomic updates for both. Currently, if `SyncedSeq`s update fails, it mainly impacts GC efficiency without causing major issues. We need to consider implementing a correction logic to remove SyncedSeq in the future.

B. We're currently updating `SyncedSeq` one by one. This approach is similar to n+1 query problem. We need to investigate if we can optimize this process by using a single query in the future.

Reply:
I'll add a comment.